### ISSUE TITLE
Arabic Numbers 

### ISSUE THREAD

**Issue by ahmed-tea** [tesseract-ocr/tesseract#1193]
### Environment

* **Tesseract Version**: Current main repository (4.00.00alpha)
* **Platform**: Windows7 32-bit

### Current Behavior:
Its recognize Arabic Characters and can not recognize Arabic numbers (ارقام عربى 0123456789)
I tried tessdata, tessdata_best, and tessdata_fast

**Comment by amitdo** [comment]
Did you try Arabic.traineddata?

**Comment by ahmed-tea** [comment]
@amitdo yes

**Comment by ahmed-tea** [comment]
It is recognize the characters (80%) and it does not recognize the numbers inside the red rectangle (the original without red rectangle )    
![my-national-identity-card-1-728](https://user-images.githubusercontent.com/32733104/32214667-4bf382c8-be28-11e7-9dc5-f415a7a4a02e.jpg)

I tried other pics with numbers only and i got no numbers
![arabnum](https://user-images.githubusercontent.com/32733104/32214946-29824b06-be29-11e7-9858-a5649222e41b.jpg)
![page0001](https://user-images.githubusercontent.com/32733104/32214999-5102d146-be29-11e7-8381-354b6ecebccf.jpg)

**Comment by Shreeshrii** [comment]
@theraysmith has not updated the repositories with changes to handle all these issues. Hence, you should not expect them to be fixed.

**Comment by ahmed-tea** [comment]
@Shreeshrii @theraysmith Is there a changes handle all these issues but the repositories did not update yet or there is no fix ?

**Comment by Shreeshrii** [comment]
I think Ray was planning to do new training to handle all these cases. But there has been no update from him since then. Based on past patterns, I would guess that he will make some updates to project before year end!

**Comment by ahmed-tea** [comment]
### Definition
* **AEN** Arabic Eastern Numbers {ِ123456789}
* **AWN** Arabic Western Numbers {0123456789}

I generated an experimental data file for recognaize **AEN Only**
The output of Tesseract OCR will be in the form of **AWN**

[https://github.com/ahmed-tea/tessdata_Arabic_Numbers](url)

@Shreeshrii @theraysmith

**Comment by Shreeshrii** [comment]
Thanks for sharing the traineddata. Please let us know the succeed rate of OCR when using it.

Do you combine it with Arabic traineddata to get correct text plus Arabic numbers using -l Ara+

**Comment by ahmed-tea** [comment]
The succeed rate for the pics above 100% (numbers only) but it depends on the pic quality in general

Combining with ara
****- current tesseract main repository** :** give an error (mgr->GetComponent(TESSDATA_INTTEMP, &fp):Error:Assert failed:in file classify\adaptmatch.cpp, line 537)
**- tesseract build by UB Mannheim :** give numbers only
**- best and fast (ara and Arabic) :** not applicable because they used for LSTM only so it give numbers only 

@Shreeshrii

**Comment by Fahad-Alsaidi** [comment]
@Shreeshrii  sorry for the question, how to combine the new tessdata_Arabic_Numbers with the current one?
I copied ara_number.traineddata into tessdata dir then I use this command:
`tesseract -l ara_number+ara  image.tif out.txt`

but doesn't work

**Comment by ahmed-tea** [comment]
@Fahad-Alsaidi You can't combine it with ara

**Comment by ahmed-tea** [comment]
@raminas81 
I think the error because it works for OEM_TESSERACT_ONLY (The old engin)
It can't combine it with ara.traindata

**Comment by AbdelsalamHaa** [comment]
Hi, I'm trying to recognize Arabic number using tesseract 3.04. The results using https://github.com/tesseract-ocr/tessdata/tree/3.04.00 train data from here with the cube files of course are very random and most of the recognize digits are wrong, is there any other traineddata file to use for only numbers, in tesseract 3.04. 


one more thing and i would be very great full , if i want to include a white List for Arabib recognition how this can be done ? 
when i use English recognition i done it as below 
![image](https://user-images.githubusercontent.com/35866217/43436023-7c444fc6-9437-11e8-9b61-b4b6597fcfc9.png)

thank you so much

**Comment by AbdelsalamHaa** [comment]
@ahmed-tea Hi , i have used your Arabic trained file for tesseract 4 and it's very good. I'm trying to do the same file but for tesseract 3.04, i could do it but the results are return in arabic as well not like your case where the numbers are return in English. I want my results to be return in English coz there's a lot of flips between the numbers order due to the language start from right to left. 
i hope you can help in this thank you so much in advance

**Comment by ahmed-tea** [comment]
@AbdelsalamHaa 
Use jTessBoxEditor [https://github.com/nguyenq/jTessBoxEditor](url) by  @nguyenq
After Box Generating and before training readjust the char corresponding to each box
The tool will not accept to enter Arabic numbers so you have to enter the English number
The OCR will read the Arabic number but the output will be English number


![](http://vietocr.sourceforge.net/jTessBoxEditor.png)

**Comment by nguyenq** [comment]
You can use your Arabic input method to enter Arabic digits, or use the built-in conversion tool. At **Character** textbox, e.g., enter `U+0668` and click the adjacent button twice or press Enter key.

**Comment by ahmed-tea** [comment]
@AbdelsalamHaa 
1- Make an Arabic Numbers jpg image 
2- In Trainer Tab select the jpg image for training data
3- Set language with the name you want for the tesseract data file
4- Select Make Box File only then run
5- In Box Editor open the jpg image
6- For each box in the image you will find corresponding character in column char (it will be wrong character)
7- Readjust each char with respect to each box (it will not accept Arabic numbers so you had to enter English numbers )
8- Save
9- Go to Trainer Tab and select Train with Existing Box and run

@nguyenq 
I tried your method
The output of OCR reorganization is the Unicode not the number

**Comment by WaelKamel116** [comment]
@Ahmed-tea
Thanks for sharing the training file.
I’ve downloaded it but did not know how to add to tesseract training files
Can you share any guide ?

**Comment by zdenop** [comment]
@ahmed-tea : is this issue solved?

**Comment by salemalbadawi** [comment]
@Shreeshrii hello I have some questions:

1. what is the best tool to train the engine some language?
2. is there a minimum size for the training dataset or image?

**Comment by salemalbadawi** [comment]
we face a problem when we  train the OCR on Indian numbers ( ١٢٣٤٥٦٧٨٩٠  ) 
also, we get a bad result when we try to read an image with a mix of Arabic and Indian numbers paragraph 
Any suggestions? 

@Shreeshrii

**Comment by Shreeshrii** [comment]
I have tried different types of fine tuning for adding the numbers but have
not had much success. I think that the open source tesseract is missing
some key component related to Arabic.
We will have to wait till @theraysmith or @jbreiden can investigate and fix
this.

**Comment by m0beenali** [comment]
@ahmed-tea @Shreeshrii Hi i'm having an issue using ahmed's traineddata. It's working fine but i **also need to recognize the slashes i.e. " / ".** I am trying to **read date from an image**, the date is written in arabic. So,

**1. How can i create a trained data file?
2. Is there any way to append or combine a traineddata file containing " / " only with ahmed's traineddata file?**

**Comment by Shreeshrii** [comment]
Reports with latest versions:

Arabic-Indic numbers incorrectly recognized #2864

Some Arabic-Indic numbers are being reversed #2897

**Comment by Shima-Mohammed** [comment]
@ahmed-tea  
thank you for sharing the training file 
did you solve the issue ?

**Comment by BasmaFahmy** [comment]
do you have any way to combine two trained files ara and ara_number?
thanks in advance

**Comment by BasmaFahmy** [comment]
https://github.com/Shreeshrii/tessdata_arabic 
this link may help, it helped me a lot.

**Comment by sam-kurdi** [comment]
is there any Indic Arabic dataset for training tesseract ?

**Comment by MahmoudMabrok** [comment]
> ### Definition
> 
>     * **AEN** Arabic Eastern Numbers {ِ123456789}
> 
>     * **AWN** Arabic Western Numbers {0123456789}
> 
> 
> I generated an experimental data file for recognaize **AEN Only**
> The output of Tesseract OCR will be in the form of **AWN**
> 
> [https://github.com/ahmed-tea/tessdata_Arabic_Numbers](url)
> 
> @Shreeshrii @theraysmith

it has a wrong link 
text of link is correct but url embded in link is invalid

**Comment by jishakrishnan** [comment]
https://github.com/jishakrishnan/pytrsseract-arabic

**Comment by jishakrishnan** [comment]
https://github.com/jishakrishnan/pytrsseract-arabic - try this out

**Comment by engahmed1190** [comment]
Hi @Shreeshrii . 

![e6d835cf894cfa4a](https://user-images.githubusercontent.com/25582838/143444012-04d9f4a2-35d5-4c51-9530-01f36e7d159e.jpg)
I have this example the date `تار يخ السداد` appears like this 

```
تار يخ السداد : 48./١./١٠؟١٠؟‏
```

Any suggestion , Thanks

**Comment by wolfassi123** [comment]
@engahmed1190  Did you manage to solve the issue cocnerning the date?

**Comment by engahmed1190** [comment]
I had to train the arabic number model on this format but still not reliable enough

**Comment by Shreeshrii** [comment]
I have not had any success in training combined Arabic text and numbers. The reason I think is that Arabic text is RTL, Arabic numbers are treated as LTR, and in training text there are sometimes unicode control characters indicating RTL and LTR. 
If separate Arabic text and Arabic number traineddata work well in recognition, that might be the way to go. I haven't tried that.

**Comment by wolfassi123** [comment]
@ Yes I have been adding several lines of my own data which includes arabic numbers and dates, and I always end up getting "Compute CTC targets failed".
I get that this issue is because I have bad training data (It's just numbers and dates but Tesseract seems to have issues with it) and it consists a reasonable percentage of my training data.
I am gonna have to train a seperate dataset for dates and ID numbers then, and one for the text.
How should I proceed to make my own training data concerning numbers and dates?

**Comment by Shreeshrii** [comment]
>I had to train the arabic number model on this format but still not reliable enough

If you can share your training process as well as the arabic numbers traineddata, if will benefit a number of users.

**Comment by engahmed1190** [comment]
I have used this [ traineddata](https://github.com/ahmed-tea/tessdata_Arabic_Numbers) and used date in this format 

![١٩٥٥%٠١%٠٢](https://user-images.githubusercontent.com/25582838/156028309-45c3f49b-ea7e-4dee-a574-43fecae202ab.jpg)
 
Still have problems in converting it to tesstrain format using https://github.com/Shreeshrii/tesstrain

I have uploaded a sample of the training data with a ground truth label [here](https://drive.google.com/file/d/1ZxU-YlUDq3vbh2I1Eipqqqx7BAAlYLWy/view?usp=sharing)

I had the same idea about separating Arabic Numbers from Arabic Text and using two traineddata instead of one but the accuracy is very low

**Comment by ramipro11** [comment]
hello 
how i can resize image dynamically for reading arabic number in tesseract

**Comment by ramipro11** [comment]
@ahmed-tea

**Comment by MostafaAbdElRasoul** [comment]
@ahmed-tea I can't find any way to combine Arabic characters and Arabic numbers in one traindata file
I need to scan a file has both Arabic characters and numbers any one can help  ?

**Comment by baraahk** [comment]
Dears,
i Have issue with numbering when OCR with C#
the result was
١‏ التخطيط للمشروع ‎١‏ تحدید الاهداف ‎١‏ وضع الجدول الزمني ‎١‏ . الجدول الزمني ۲ ,. الجدول الزمني ‎.٢‏ التخطيط للمشروع ١ر۲‏ تحدید الاهداف ۲ . وضع الجدول الزمني ‎.۲,۲١‏ الجدول الزمني ۲ . الجدول الزمني ۳ ,. الجدول الزمني ۳ الجدول الزمني

and check the file as attached 

[testarabic number (1).pdf](https://github.com/user-attachments/files/24014128/testarabic.number.1.pdf)


i need your help , it is very urgent case
